### PR TITLE
Integrate Perfetto as backend for ScopedProfiling

### LINKED ISSUE #1224: Integrate Perfetto for optional CPU profiling
It looks like [Perfetto](https://github.com/google/perfetto), the Google performance/tracing tool, could be useful to integrate into Celeritas to measure performance. It might be straightforward to wrap their [tracing SDK](https://perfetto.dev/docs/instrumentation/tracing-sdk) underneath `ScopedProfiling`.

### PR DESCRIPTION
**Pull request by esseivaju**
Add Perfetto as an external dependency. It can be used as a backend for the existing `ScopedProfiling` when CUDA and HIP are disabled. The two main additions are:
*  ```celeritas::trace_counter``` to record the evolution of numeric values over time
* ```TracingSession``` as a wrapper around a perfetto session so that only corecel has to depend on Perfetto. A ```TracingSession``` needs to be created to record data when using Perfetto.

Since Perfetto system-level tracing requires Linux and root privileges, it can't be used on most clusters, so this also provides a Dockerfile for a development container that can be used on non-Linux hosts.

### CODE REVIEW

File: cmake/FindPerfetto.cmake
Comment (sethrj) [inline cmake/FindPerfetto.cmake]: I think if we're going to add a library in our own namespace it shouldn't be in "find package"... maybe we should make an externals for like with g4vg: https://github.com/celeritas-project/g4vg/blob/main/external/CMakeLists.txt

File: src/corecel/CMakeLists.txt
Comment (sethrj) [inline src/corecel/CMakeLists.txt]: Ideally the includes would be part of the `perfetto` target and would also be `PRIVATE`...

File: app/celer-sim/celer-sim.cc
Comment (sethrj) [inline app/celer-sim/celer-sim.cc]: I assume this is more or less debugging/playing around for now? I think it would be sensible for all this code to go in ScopedProfiling... and perhaps even the initialization could go inside the `ScopedProfiling::use_profiling` IIFE ?

**Review by sethrj — COMMENTED** [review]
Don't worry about comments until you have a sense that Perfetto could actually be useful and interesting for c++ development/profiling.

File: src/corecel/CMakeLists.txt
Comment (esseivaju) [reply src/corecel/CMakeLists.txt]: This should be fixed now

**Comment by esseivaju** [comment]
I'm not sure why doxygen isn't happy with the documentation

File: src/corecel/sys/Counter.hh
Comment (sethrj) [inline src/corecel/sys/Counter.hh]:
```suggestion
//---------------------------------------------------------------------------//
/*!
 * Simple tracing counter.
 * Records a named value at the current timestamp which
```

File: src/corecel/sys/Counter.hh
Comment (sethrj) [inline src/corecel/sys/Counter.hh]:
```suggestion
 * \tparam T Arithmetic counter type
```

File: cmake/FindPerfetto.cmake
Comment (sethrj) [reply cmake/FindPerfetto.cmake]: This is still true. Also if we're making a `STATIC` library doesn't it have to be after our global "everything is relocatable" setup?

File: scripts/docker/interactive/build.sh
Comment (sethrj) [inline scripts/docker/interactive/build.sh]: I'm not familiar with `shellcheck`, is that something we should put in our CI? Also should we `set -e` and use `sh` instead of forcing `bash`?

File: scripts/docker/interactive/run-perfetto.sh
Comment (sethrj) [inline scripts/docker/interactive/run-perfetto.sh]: Probably better to exec?
```suggestion
exec tools/tmux -c celer.cfg -C out/linux -n
```
Same comment about bash above.

File: scripts/docker/interactive/run.sh
Comment (sethrj) [inline scripts/docker/interactive/run.sh]: Same exec comment

File: scripts/docker/interactive/run.sh
Comment (sethrj) [inline scripts/docker/interactive/run.sh]: Can't you use `while getopts`?

E.g.
```sh
OPTIND=1
while getopts "fb:p:G:" opt; do
  case "${opt}" in
    t) TAG_NAME =${OPTARG} ;;
    m) HOST_MOUNT ="${OPTARG}" ;;
    \?) echo "Invalid option: -$OPTARG" >&2; exit 1 ;;
    *) usage; exit 2; ;;
  esac
done
shift $((OPTIND-1))
```

File: src/corecel/CMakeLists.txt
Comment (sethrj) [inline src/corecel/CMakeLists.txt]: indentation

File: src/corecel/sys/Counter.hh
Comment (sethrj) [inline src/corecel/sys/Counter.hh:20]: Should there be `extern template` declarations as well?

File: src/corecel/sys/ScopedProfiling.hh
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh:22]: Add doxygen and note the env variable name

File: src/corecel/sys/ScopedProfiling.hh
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]: I really would rather have the static data and getenv/logger calls in a .cc file. Maybe change `.perfetto.cc` to `.cc` and `#if` the perfetto parts?

File: src/corecel/sys/ScopedProfiling.hh
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]: I wonder if we should also check for the existence of the environment variable to warn users?

File: src/corecel/sys/ScopedProfiling.hip.cc
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hip.cc:42]: This warning code should go back in there for the device case

File: src/corecel/sys/ScopedProfiling.perfetto.cc
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.perfetto.cc]:
```suggestion
{
//---------------------------------------------------------------------------//
/*!
```
and below

File: src/corecel/sys/TracingSession.hh
Comment (sethrj) [inline src/corecel/sys/TracingSession.hh]:
```suggestion
 * Constructors will only configure and initialize the session. It needs to
 * be started explicitly by calling \c TracingSession::start
```

File: src/corecel/sys/TracingSession.hh
Comment (sethrj) [inline src/corecel/sys/TracingSession.hh]:
```suggestion
    // Terminate the session and close open files
```

File: src/corecel/sys/TracingSession.hh
Comment (sethrj) [inline src/corecel/sys/TracingSession.hh]: Instead of declaring a second class, could you take the normal approach of providing `inline` `CELER_NOT_CONFIGURED (or null-op I guess?)` definitions?

File: src/corecel/sys/TracingSession.perfetto.cc
Comment (sethrj) [inline src/corecel/sys/TracingSession.perfetto.cc]: Could you please recheck the merge request for formatting and proper doxygen markup? There are a lot of inconsistencies with the rest of the codebase.

File: src/corecel/sys/TracingSession.perfetto.cc
Comment (sethrj) [inline src/corecel/sys/TracingSession.perfetto.cc]: Since the fd already has a default initializer to -1, I think this is far more readable:
```suggestion
    : session_{initialize_session(TracingMode::InProcess)}
{
    if (session_)
    {
        fd_ = open(filename.data(), O_RDWR | O_CREAT | O_TRUNC, 0660);
        session_->Setup(configure_session(), fd_);
```

File: src/corecel/sys/detail/TrackEvent.perfetto.hh
Comment (sethrj) [inline src/corecel/sys/detail/TrackEvent.perfetto.hh]: Doesn't this need a static, inline, or a `.cc` file?

**Review by sethrj — CHANGES_REQUESTED** [review]

File: src/corecel/sys/detail/TrackEvent.perfetto.hh
Comment (esseivaju) [reply src/corecel/sys/detail/TrackEvent.perfetto.hh]: Right, I forgot inline. I don't think static is needed though.

File: src/corecel/sys/ScopedProfiling.hh
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: We can't make that function constexpr if we want to do this here.

I updated it but now the compiler can't tell that constructing a ScopedProfiling without CUDA/HIP/Perfetto is a no-op.

File: src/corecel/sys/TracingSession.hh
Comment (esseivaju) [reply src/corecel/sys/TracingSession.hh]: I think in that case, null-op is fine, maybe add a warning here is the environment variable is set.

File: src/corecel/sys/Counter.hh
Comment (esseivaju) [reply src/corecel/sys/Counter.hh:20]: I'm not sure it's needed: the definition is not accessible from the header so we have to rely on explicit instantiation anyway or we get a linker error

File: scripts/docker/interactive/build.sh
Comment (esseivaju) [reply scripts/docker/interactive/build.sh]: I forced bash instead of sh because I'm using `BASH_SOURCE` to find the script directory but we can change it to `dirname $0` it will work in most cases. 

I like to use `shellcheck` since bash has many pitfalls 😅 I suppose it wouldn't hurt to have it in the CI.

File: cmake/FindPerfetto.cmake
Comment (esseivaju) [reply cmake/FindPerfetto.cmake]: I introduced it in our namespace because it gets compiled with Celeritas but I'm not sure that's the best option. We're handling all our dependencies using find_package, so I also went with that.

This is after the relocatable setup no? Perfetto.cc gets compiler with `-fPIC`

File: src/corecel/sys/ScopedProfiling.hh
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.hh]: Ok, done. However, I like the inline static more than the conditional compilation. What's the reason for preferring logger calls in cc? compile time?

**Comment by esseivaju** [comment]
Here's an example of application and system traces: [traces.tar.gz](https://github.com/celeritas-project/celeritas/files/15441356/traces.tar.gz)

I've noticed that on my laptop (16 cores) if I use the in-process mode and max out the CPU with 16 OpenMP threads, some data is lost. It doesn't happen when using the system mode.

File: CMakeLists.txt
Comment (esseivaju) [inline CMakeLists.txt]: can we require a newer CMake version (3.14) or is 3.12 still important to support?

**Comment by esseivaju** [comment]
@sethrj I'll update the CMake integration once #1253 is merged. Other than that it should be ready for a second look.

File: CMakeLists.txt
Comment (sethrj) [inline CMakeLists.txt]: Since it sounds like perfetto isn't ever externally installable, I think this should just be an `option(CELERIITAS_USE_Perfetto` like we (will) have for JSON.

File: CMakeLists.txt
Comment (sethrj) [inline CMakeLists.txt]: Does this show up only on the first configure or on every configure? (Also for cmake use `if(` instead of `if (` since it's a function 🙄 )

File: app/celer-sim/RunnerInputIO.json.cc
Comment (sethrj) [inline app/celer-sim/RunnerInputIO.json.cc]: Probably `LDIO_SAVE_WHEN(tracing_file, CELERITAS_USE_PERFETTO)`?

File: app/celer-sim/celer-sim.cc
Comment (sethrj) [inline app/celer-sim/celer-sim.cc]: Maybe the `tracing_file` run input should have the default parameter (if perfetto is active) or the constructor could be simplified?

File: src/corecel/sys/ScopedProfiling.cc
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.cc]: The `use_profiling` could probably be combined with the `.cuda.cc` version: we're just using a different profiler. Or maybe at this point we should just merge everything into a single file... thoughts?

File: src/corecel/sys/ScopedProfiling.hh
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hh]:
```suggestion
 * Celeritas is built with device support (CUDA/HIP)
```

File: src/corecel/sys/ScopedProfiling.hip.cc
Comment (sethrj) [inline src/corecel/sys/ScopedProfiling.hip.cc]: Possibly could also be combined into the `.cc` file

File: src/corecel/sys/TracingSession.cc
Comment (sethrj) [inline src/corecel/sys/TracingSession.cc:26]: Is there a reason not to just put this anonymous namespace inside the celeritas one like we usually do?

**Review by sethrj — COMMENTED** [review]
This is really cool, I think this would be worth at least a poster session at CHEP or the like!

File: CMakeLists.txt
Comment (sethrj) [reply CMakeLists.txt]: I'm comfortable with requiring up to 3.20 (3 year support window).

File: CMakeLists.txt
Comment (esseivaju) [reply CMakeLists.txt]: it shows every time, I removed it

File: app/celer-sim/celer-sim.cc
Comment (esseivaju) [reply app/celer-sim/celer-sim.cc]: I moved that check inside the constructor, a bit more ergonomic to use that way

File: src/corecel/sys/ScopedProfiling.cc
Comment (esseivaju) [reply src/corecel/sys/ScopedProfiling.cc]: we can definitely combine `use_profiling`. I think I still like having separate files for the other functions, it makes it easier to find the implementation you're looking for, if anything I'd like to split `use_profiling` in a separate file and remove the `#if CELERITAS_USE_PERFETTO` in the `.cc`

**Review by sethrj — APPROVED** [review]
Looks good to me! We can wait till #1253 is merged and then move the external fetching into `external/`.

File: CMakeLists.txt
Comment (esseivaju) [inline CMakeLists.txt:331]: Should we provide a FindPerfetto module here? Otherwise, find_package will never succeed since Perfetto doesn't support CMake.

Comment (sethrj) [reply CMakeLists.txt:331]: I guess we could, is it something users even have installed externally? I'm fine with `find_package` never succeeding....

**Review by sethrj — APPROVED** [review]

File: CMakeLists.txt
Comment (esseivaju) [reply CMakeLists.txt:331]: No, it shouldn't be externally installed.  Having ```FETCHCONTENT_SOURCE_DIR_PERFETTO``` use a local checkout instead of fetching it is fine. I should have added QUIET to the find_or_external call as we get a warning for not having a module and config file.

**Pull request merged**
